### ISSUE TITLE
More statistical-based metrics (FRR@FAR or FNMR@FMR)

### ISSUE THREAD

**Issue by dapladoc** [OML-Team/open-metric-learning#226]
Sometimes it might be helpful or just interesting to see some statistical-based metrics in order to get some sort of insight about what could be done to improve the model, or why the current rank-based metrics are as they are. For example, it might be interesting to track mean distances among positive and negative pairs. Even for very good looking retrieval metrics it might happen that distances among positives and negatives are too close to each other, and that could be a bad sign. There are commonly used in face recognition systems FRR@FAR (or FNMR@FMR), equal error rate, detection error tradeoff and so on. Does it make sense to add this type of metrics to oml?

**Comment by AlekseySh** [comment]
Hi, @dapladoc !

> For example, it might be interesting to track mean distances among positive and negative pairs
We already have this. Moreover, we track the distances with respect to the source of the pair (like if we take it from the memory bank or from the original batch). It helps us to see if dimension collapse has happened (which is often a situation for triplet loss).

**Comment by AlekseySh** [comment]
![Untitled](https://user-images.githubusercontent.com/20143575/202854436-d67f97f5-e924-48fa-ae9b-371f996fefcc.jpg)

**Comment by AlekseySh** [comment]
As for FRR@FAR, could you please check if the information given to the `calc_retrieval_metrics` function is enough to calculate it? The idea of having one function for metrics instead of several is that they share some heavy computational operations. Is this new metric sharing some steps with the existing precision, recall and map or it can be implemented as a separate function without overlapping logic?

**Comment by dapladoc** [comment]
I see that pos_dist and neg_dist are evaluated only for train data. I'd suggest to measure pos_dist and neg_dist for validation data too. One more measure of under/over-fitting will never be superfluous. As for FRR@FAR it seems that `calc_retrieval_metrics` has all the ingredients required, and for its evaluation one needs to have an array of all the positive distances, and an array of all negative distances, thus, it costs nothing to evaluate their mean values while evaluating FRR@FAR.

**Comment by AlekseySh** [comment]
Is it correct that  FRR@FAR is distance-sensitive (while other metrics are only order-sensitive)?

**Comment by dapladoc** [comment]
For me it looks like FRR@FAR is also just order-sensitive by construction. Basically, in order to evaluate for example FRR@FAR-1.e-4 (i.e. FRR value when FAR is 1.e-4) you need to
1. Find such threshold, that FAR is equal to 1.e-4. It could be done by using
```python
threshold = torch.quantile(neg_dist, desired_far_value)
```
2. Evaluate the ratio of positive samples that are above the `threshold`
```
frr_at_far = (pos_dist >= threshold).sum() / len(pos_dist)
```
So concrete distance values don't matter.

**Comment by AlekseySh** [comment]
@dapladoc got it.

What shall we do here? As far as I understood, originally you wanted to suggest some distance-sensitive metrics (in addition to those that we have for the training). We've realised, that the last one is not from this type of metric. 

So, you may come up with other distance-specific metrics and we will think if we want to add them to the repo.

Or, you can implement FRR@FAR (if I got correctly that you gonna implement it?). This option is also fine since it's a well-known metric among people with an identification/verification background.

What are your thoughts regarding that?

**Comment by AlekseySh** [comment]
From my side, I can mention the way how authors of [UNDERSTANDING DIMENSIONAL COLLAPSE IN CONTRASTIVE SELF-SUPERVISED LEARNING](https://arxiv.org/pdf/2110.09348.pdf) built an approach to check if dimension collapse of feature vectors has happened. They check the spectrum of singular values and the idea was that a good model does not produce a lot of singular values close to 0. It's not a 1-value metric out of the box, but we can build the metric based on this spectrum (for example, take a percentile of non-zero singular values).

You can find the paper for more details if it sounds interesting. It does not work with distances directly, but still.

**Comment by dapladoc** [comment]
> As far as I understood, originally you wanted to suggest some distance-sensitive metrics

The means of so called positive and negative distances are examples of these metrics. Another one that I have in mind is earth mover's distance (or Wasserstein distance) between distributions of positive and negative distances. There is an implementation of it in scipy ([link](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html)), but as for pytorch, it is a trickier question. There are some implementations of it, but I don't know whether they are trustful or acceptable in terms of computational resources.

Is it correct that only the pytorch-based methods are valid for metrics computation? Or it is ok to convert the distance matrix to numpy and apply numpy-based functions to it?

**Comment by dapladoc** [comment]
> From my side, I can mention the way how authors of [UNDERSTANDING DIMENSIONAL COLLAPSE IN CONTRASTIVE SELF-SUPERVISED LEARNING](https://arxiv.org/pdf/2110.09348.pdf) built an approach to check if dimension collapse of feature vectors has happened. They check the spectrum of singular values and the idea was that a good model does not produce a lot of singular values close to 0. It's not a 1-value metric out of the box, but we can build the metric based on this spectrum (for example, take a percentile of non-zero singular values).
> 
> You can find the paper for more details if it sounds interesting. It does not work with distances directly, but still.

You already have an issue related to PCA. That's basically the same.

**Comment by dapladoc** [comment]
As for FNMR@FMR there is an issue related to `torch.quantile` ([Quantile is limited to 16 million elements and have poor performance.](https://github.com/pytorch/pytorch/issues/64947)). In that thread there was suggested a c++ extension that implements unrestricted version of `quantile` function ([link](https://github.com/pytorch/pytorch/issues/64947#issuecomment-922343374)). I tried it and it seems to work. But there could be issues with its compilation on various platforms. I didn't work with `torch.utils.cpp_extension.load_inline` and I don't know its drawbacks.

So, I see 2 ways how FRR@FAR can be implemented
1. By using the proposed c++ extension of `quantile` ([link](https://github.com/pytorch/pytorch/issues/64947#issuecomment-922343374)).
Pros:
* easy to use
* less code for FRR@FAR evaluation itself
* exact solution.
Cons:
* the code needs to be maintained by someone in case of pytorch version update
* there could be problems with compilation on various platforms.
2. The required threshold could be evaluated approximately by using cumsum of histogram of negative distances. We can find two bins in the histogram, that contains the desired amount of negative distances values, and interpolate between them to find approximation of the threshold.
Pros:
* no auxiliary dependencies.
Cons:
* the estimated FRR@FAR will be not exact, and the exactness depends on the number of bins in the histogram of the negative distances.

**Comment by AlekseySh** [comment]
> > As far as I understood, originally you wanted to suggest some distance-sensitive metrics
> 
> The means of so called positive and negative distances are examples of these metrics. Another one that I have in mind is earth mover's distance (or Wasserstein distance) between distributions of positive and negative distances. There is an implementation of it in scipy ([link](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html)), but as for pytorch, it is a trickier question. There are some implementations of it, but I don't know whether they are trustful or acceptable in terms of computational resources.
> 
> Is it correct that only the pytorch-based methods are valid for metrics computation? Or it is ok to convert the distance matrix to numpy and apply numpy-based functions to it?
> 
> UPD. By the way, it could be very interesting to track EMD not only between positive and negative distances, but between themselves on different validation steps, i.e. `EMD(pos_dist[t], pos_dist[t - 1])`.

I am unsure why distance distributions for positive and negative values should be similar. Even if we track this I don't know how to interpret the results. 

As for metric dimension collapse. Yes, it's similar to PCA. We can track the amount of non-trivial singular values as a metric. It's an optional idea. We don't need to discuss it here.

**Comment by AlekseySh** [comment]
As for FNMR@FMR, I don't like both ideas.
Do we have any problems if we allow ourselves to use NumPy?

**Comment by dapladoc** [comment]
> As for FNMR@FMR, I don't like both ideas.
> Do we have any problems if we allow ourselves to use NumPy?

With NumPy there should be no problems, except maybe the performance. 

> I am unsure why distance distributions for positive and negative values should be similar or not similar.

They should be not similar. The positive distances distribution should be concentrated near 0, while the negative distances distribution should be as far from 0 as possible. So, the bigger EMD for positive-vs-negative distances distributions the better. As for EMD for positive-vs-positive or negative-vs-negative it should show how much the distributions change over time in general during training. For instance, it is possible that mean values for positive and negative distances don't change between iterations, but EMD could increase, because, say, positive distribution changes its shape from the bell shape to a skewed-bell-shape ditribution.

Here is an example. The green histograms are like distributions of positive distances, and the red one are like distributions of negative distances. On the both plots the mean values of green histograms are 1, and mean values for the red histograms are 6. And here the EMD distances between them:
for the left plot EMD between positives and negatives is 4.978295629872207
for the right plot EMD between positives and negatives is 5.015395329150344
EMD between positives is 0.2539768070093981
EMD between negatives is 1.5615436701286642

![example](https://user-images.githubusercontent.com/6040221/203086446-927713fb-67e2-4373-b1fa-4f0f5403090a.png)

As for EMD between positives and negatives I'm surprised that the distances on both plots are very similar, but having all four numbers and mean values we actually can imagine what is going on with the distributions.

**Comment by AlekseySh** [comment]
> They should be not similar.

Sorry, I confused you. I meant their shapes. Of course, their average values have to be different (smaller ones for pos, higher ones for neg), but the average distances are already tracked in OML. That is why I thought that you want to track only the changes in distribution shapes (after subtraction of the mean values of the distributions).

**Comment by AlekseySh** [comment]
> for the left plot EMD between positives and negatives is 4.978295629872207
> for the right plot EMD between positives and negatives is 5.015395329150344
> I'm surprised that the distances on both plots are very similar

I agree and am also surprised. That is why my suggestion was to log the distributions directly, rather than logging some metric like EMD and then trying to do reverse engineering. What do you think?

**Comment by AlekseySh** [comment]
> With NumPy there should be no problems, except maybe the performance.

If the overhead of transforming the distances matrix into NumPy is not huge, we can go for NumPy. 

PS. If there is a way to avoid using `quantile` in torch? Probably we can find some alternative implementations as references

**Comment by AlekseySh** [comment]
As for FNMR@FMR:
First, I suggest checking the performance of the torch `quantile`. How bad is it? As far as I remember there are 12k queries and 13k galleries in the InShop dataset (roughly 1.5 times more in SOP). Since we perform this operation only once during the epoch, it may be probably okay.

**Comment by dapladoc** [comment]
Buy the way, is'n it better to introduce functions like `calc_map`, `calc_cmc` etc, and the metrics code will look like
<details>
<summary>The code</summary>

```python
def calc_retrieval_metrics(
    distances: torch.Tensor,
    mask_gt: torch.Tensor,
    mask_to_ignore: Optional[torch.Tensor] = None,
    cmc_top_k: Tuple[int, ...] = (5,),
    precision_top_k: Tuple[int, ...] = (5,),
    map_top_k: Tuple[int, ...] = (5,),
    fmr_vals: Tuple[float, ...] = (1.0e-3,),
    reduce: bool = True,
    check_dataset_validity: bool = False,
) -> TMetricsDict:
...
    for k_cmc, k_cmc_show in zip(cmc_top_k_clipped, cmc_top_k):
        metrics["cmc"][k_cmc_show] = calc_cmc(gt_tops, k_cmc)

    for k_precision, k_precision_show in zip(precision_top_k_clipped, precision_top_k):
        metrics["precision"][k_precision_show] = calc_precision(mask_gt, gt_tops, k_precision)

    for k_map, k_map_show in zip(map_top_k_clipped, map_top_k):
        metrics["map"][k_map_show] = calc_mean_ap(mask_gt, gt_tops, k_map)
    ...
    return metrics


def calc_cmc(gt_tops, k):
   ...


def calc_precision(mask_gt, gt_tops, k):
   ...


def calc_mean_ap(mask_gt, gt_tops, k):
...



```
</details>

And in test scripts each `calc_...` can be tested independently, in addition to the overall `calc_retrieval_metrics` test.

**Comment by AlekseySh** [comment]
Your version of the code is better, if you want you can work on it (in a dedicated PR, which includes only refactoring, without bringing the new functionality), but I don't see this has a high priority.
